File: Programmers/cpp/LEVEL_3_단어_변환.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

bool check(string a, string b) {
	int diff = 0;

	for (int i = 0; i < a.length(); i++) {
		if (a[i] != b[i]) {
			diff++;
		}
	}

	if (diff == 1) {
		return true;
	}
	return false;
}

int solution(string begin, string target, vector<string> words) {
	int answer = 0;
	vector<bool> visit(words.size(), false);

	queue<pair<string, int>> q;
	q.push({ begin, 0 });

	while (!q.empty()) {
		string word = q.front().first;
		int cnt = q.front().second;
		q.pop();

		if (word == target) {
			answer = cnt;
			break;
		}

		for (int i = 0; i < words.size(); i++) {
			if (visit[i] == false && check(word, words[i])) {
				visit[i] = true;
				q.push({ words[i], cnt + 1 });
			}
		}
	}

	return answer;
}

File: Programmers/java/LEVEL_3_단어_변환.java
import java.util.*;

class Solution {
    private boolean check(String a, String b) {
        int diff = 0;

        for (int i = 0; i < a.length(); i++) {
            if (a.charAt(i) != b.charAt(i)) {
                diff++;
            }
        }

        if (diff == 1) {
            return true;
        }
        return false;
    }

    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        boolean[] visit = new boolean[words.length];

        Queue<Pair> q = new LinkedList<>();
        Pair pair = new Pair(begin, 0);
        q.add(pair);

        while (!q.isEmpty()) {
            pair = q.poll();
            String word = pair.getWord();
            int cnt = pair.getCount();

            if (word.equals(target)) {
                answer = cnt;
                break;
            }

            for (int i = 0; i < words.length; i++) {
                if (visit[i] == false && check(word, words[i])) {
                    visit[i] = true;
                    pair = new Pair(words[i], cnt + 1);
                    q.add(pair);
                }
            }
        }

        return answer;
    }
}

class Pair {
    String word;
    int count;

    Pair(String word, int count) {
        this.word = word;
        this.count = count;
    }

    String getWord() {
        return word;
    }

    int getCount() {
        return count;
    }
}